client/jetpack-connect/signup.js

```diff
@@ -212,7 +262,128 @@ export class JetpackSignup extends Component {
+					'Your account will enable you to start using the features and benefits offered by %(pluginName)s',
+					{
+						args: {
+							pluginName: wooDna.name( translate ),
+						},
+					}
+				);
+				pageTitle = translate( 'Login to WordPress.com' );
+				footerLinks.push(
+					<LoggedOutFormLinkItem key="signup" onClick={ this.showWooDnaSignupView }>
+						{ this.props.translate( 'Create a new account' ) }
+					</LoggedOutFormLinkItem>
+				);
+				footerLinks.push(
+					<LoggedOutFormLinkItem
+						key="lostpassword"
+						href={ addQueryArgs(
+							{ action: 'lostpassword' },
+							login( { locale: this.props.locale } )
+						) }
+					>
+						{ this.props.translate( 'Lost your password?' ) }
+					</LoggedOutFormLinkItem>
+				);
+			} else {
+				header = wooDna.name( translate );
+				subHeader = translate( 'Enter your email address to get started' );
+				pageTitle = translate( 'Connect' );
+			}
+			content = (
+				<LoginBlock
+					locale={ this.props.locale }
+					footer={ this.renderWooDnaFooter( footerLinks ) }
+					userEmail={ email }
+					socialConnect={ loginSocialConnect }
+					twoFactorAuthType={ loginTwoFactorAuthType }
+					onTwoFactorRequested={ ( authType ) =>
+						this.setState( { loginTwoFactorAuthType: authType } )
+					}
+					onSocialConnectStart={ () => this.setState( { loginSocialConnect: true } ) }
+				/>
+			);
+		} else {
+			// Woo DNA sign-up form
+			header = wooDna.name( translate );
+			subHeader = translate( 'Create an account' );
+			pageTitle = translate( 'Create a WordPress.com account' );
+			footerLinks.push(
+				<LoggedOutFormLinkItem key="login" onClick={ () => this.showWooDnaLoginView() }>
+					{ this.props.translate( 'Log in with an existing WordPress.com account' ) }
+				</LoggedOutFormLinkItem>
+			);
+			content = (
+				<SignupForm
+					disabled={ isCreatingAccount }
+					email={ includes( email, '@' ) ? email : '' }
+					footerLink={ this.renderWooDnaFooter( footerLinks ) }
+					handleLogin={ this.showWooDnaLoginView }
+					handleSocialResponse={ this.handleSocialResponse }
+					isSocialSignupEnabled={ isEnabled( 'signup/social' ) }
+					locale={ this.props.locale }
+					redirectToAfterLoginUrl={ addQueryArgs( { auth_approved: true }, window.location.href ) }
+					submitButtonText={ this.props.translate( 'Create your account' ) }
+					submitForm={ this.handleSubmitSignup }
+					submitting={ isCreatingAccount }
+					suggestedUsername={ includes( email, '@' ) ? '' : email }
+				/>
+			);
+		}
+
+		return (
+			<MainWrapper wooDna={ wooDna } pageTitle={ wooDna.name( translate ) + ' — ' + pageTitle }>
+				<div className="jetpack-connect__authorize-form">
+					{ this.renderLocaleSuggestions() }
+					<FormattedHeader headerText={ header } subHeaderText={ subHeader } />
+					{ content }
+					{ this.renderLoginUser() }
+				</div>
+			</MainWrapper>
+		);
+	}
+
 	render() {
+		if ( this.getWooDnaConfig() ) {
+			return this.renderWooDna();
+		}
```

Comment:
> it will be hard to extract shared logic

There is a lot of shared logic between those 2 "components". They're just doing the same tasks in slightly different ways. I think it would be counter-productive to separate them because then we would need to keep them "in sync", which won't happen.

Reply:
> As for keeping it in sync we have to do it anyway for render and renderWooDna methods.

Yeah, I was more worried about duplicating the rest of the logic. But if it can be extracted with react hooks, it should be possible to keep it relatively clean.

> That's why I think it won't be a simple refactoring and should be handled separately.

I've created https://github.com/Automattic/wp-calypso/issues/43187 to tackle this after the deadline.